src/engine/controls/macrocontrol.cpp

```diff
@@ -0,0 +1,333 @@
+#include "macrocontrol.h"
+
+#include <QRegExp>
+
+#include "moc_macrocontrol.cpp"
+#include "track/track.h"
+
+namespace {
+constexpr uint kRecordingTimerInterval = 100;
+constexpr size_t kRecordingQueueSize = kRecordingTimerInterval / 10;
+
+const QString recordingSuffix = QStringLiteral(" [Recording]");
+const QString interruptedSuffix = QStringLiteral(" [Interrupted]");
+const QRegExp kAutoGeneratedLabelRegExp(QStringLiteral("[0-9. ]+(\\[\\w*\\])?"));
+} // namespace
+
+ConfigKey MacroControl::getConfigKey(const QString& name) {
+    return ConfigKey(m_group, QStringLiteral("macro_%1_%2").arg(m_slot).arg(name));
+}
+
+MacroControl::MacroControl(const QString& group, UserSettingsPointer pConfig, int slot)
+        : EngineControl(group, pConfig),
+          m_slot(slot),
+          m_queuedJumpTarget(-1),
+          m_recordedActions(kRecordingQueueSize),
+          m_iNextAction(0),
```

Comment:
```suggestion
          m_nextActionIndex(0),
```
(also in the corresponding code)